app/lib/pre_assembly/from_bundle/structural_builder.rb

```diff
@@ -0,0 +1,120 @@
+            content_md_creation_style: content_md_creation_style, reading_order: reading_order,
+            all_files_public: all_files_public).build
+      end
+
+      def initialize(filesets:, cocina_object:, common_path:, content_md_creation_style:, reading_order:, all_files_public:)
+        @filesets = filesets
+        @cocina_object = cocina_object
+        @common_path = common_path
+        @content_md_creation_style = content_md_creation_style
+        @reading_order = reading_order
+        @all_files_public = all_files_public
+      end
+      # rubocop:enable Metrics/ParameterLists
+
+      attr_reader :content_md_creation_style, :filesets, :cocina_object, :common_path, :reading_order, :all_files_public
+
+      # rubocop:disable Metrics/AbcSize
+      # rubocop:disable Metrics/MethodLength
+      def build
+        # a counter to use when creating auto-labels for resources, with incremenets for each type
+        resource_type_counters = Hash.new(0)
+
+        # rubocop:disable Metrics/BlockLength
+        cocina_filesets = filesets.map.with_index(1) do |fileset, sequence|
+          resource_type_counters[fileset.resource_type_description] += 1 # each resource type description gets its own incrementing counter
+          # create a generic resource label if needed
+          default_label = "#{fileset.resource_type_description.capitalize} #{resource_type_counters[fileset.resource_type_description]}"
+          # but if one of the files has a label, use it instead
+          resource_label = fileset.label_from_file(default: default_label)
+          contained_files = fileset.files.map do |obj| # iterate over all the files in a resource
+            file_id = obj.file_id(common_path: common_path)
+            file_attributes = {
+              type: 'https://cocina.sul.stanford.edu/models/file',
+              externalIdentifier: "https://cocina.sul.stanford.edu/file/#{SecureRandom.uuid}",
+              version: version,
+              label: file_id,
+              filename: file_id,
+              hasMessageDigests: message_digests(obj),
+              hasMimeType: obj.mimetype,
+              administrative: administrative(obj),
+              access: file_access
+            }
+            Cocina::Models::File.new(file_attributes)
+          end
+
+          fs_attributes = {
+            label: resource_label,
+            version: version,
+            externalIdentifier: "#{external_identifier.delete_prefix('druid:')}_#{sequence}",
+            type: file_set_type(fileset.resource_type_description),
+            structural: { contains: contained_files }
+          }
+
+          Cocina::Models::FileSet.new(fs_attributes)
+        end
+        # rubocop:enable Metrics/BlockLength
+
+        attributes = { contains: cocina_filesets }
+        attributes[:hasMemberOrders] = [{ viewingDirection: reading_order }] if content_md_creation_style == :simple_book
+
+        Cocina::Models::DROStructural.new(attributes)
```

Comment:
This should merge with the existing structural in DSA, in case a collection has been set, so as not to overwrite it.